File: src/apps/scala/slate-examples/src/main/scala/slate/examples/Example_DateTime.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slate.examples


//<doc:import_required>
import slate.common.{Ensure, DateTime}
import slate.common.results.ResultSupportIn
//</doc:import_required>

//<doc:import_examples>
import java.time._
import slate.core.cmds.Cmd
//</doc:import_examples>


class Example_DateTime extends Cmd("types") with ResultSupportIn {

  override protected def executeInternal(args: Any): AnyRef = {


    //<doc:examples>
    // Case 1. local date time by default
    val dt = DateTime.now()
    println("slate Current DateTime: " + dt)

    // Case 2. utc
    val dateOnly = dt.atUtc()
    println("dateOnly : " + dateOnly)

    // Case 3. Date parts
    println( "year  : " + dt.year    )
    println( "month : " + dt.month   )
    println( "day   : " + dt.day     )
    println( "hour  : " + dt.hours   )
    println( "mins  : " + dt.minutes )
    println( "secs  : " + dt.day     )

    // Case 4. add time
    val dt1 = new DateTime(2016, 7, 22, 8, 30, 30)
    println( dt1.addSeconds(1).toString )
    println( dt1.addMinutes(1).toString    )
    println( dt1.addHours(1).toString   )
    println( dt1.addDays(1).toString    )
    println( dt1.addMonths(1).toString  )
    println( dt1.addYears(1).toString   )

    // Case 5. Compare local
    ensureTrue(dt1, "> ", dt1.addHours(-1), dt1 >  dt1.addHours(-1) )
    ensureTrue(dt1, ">=", dt1.addHours(-1), dt1 >= dt1.addHours(-1) )
    ensureTrue(dt1, ">=", dt1.addHours( 0), dt1 >= dt1.addHours( 0) )
    ensureTrue(dt1, "< ", dt1.addHours( 1), dt1 <  dt1.addHours( 1) )
    ensureTrue(dt1, "<=", dt1.addHours( 1), dt1 <= dt1.addHours( 1) )
    ensureTrue(dt1, "<=", dt1.addHours( 0), dt1 <= dt1.addHours( 0) )
    ensureTrue(dt1, "==", dt1.addHours( 0), dt1 == dt1.addHours( 0) )
    ensureTrue(dt1, "!=", dt1.addHours( 2), dt1 != dt1.addHours( 2) )

    // Case 6. Compare local with utc
    ensureTrue(dt1, "> ", dt1.addHours(-1), dt1 >  dt1.addHours(-1).atUtc() )
    ensureTrue(dt1, ">=", dt1.addHours(-1), dt1 >= dt1.addHours(-1).atUtc() )
    ensureTrue(dt1, ">=", dt1.addHours( 0), dt1 >= dt1.addHours( 0).atUtc() )
    ensureTrue(dt1, "< ", dt1.addHours( 1), dt1 <  dt1.addHours( 1).atUtc() )
    ensureTrue(dt1, "<=", dt1.addHours( 1), dt1 <= dt1.addHours( 1).atUtc() )
    ensureTrue(dt1, "<=", dt1.addHours( 0), dt1 <= dt1.addHours( 0).atUtc() )
    ensureTrue(dt1, "==", dt1.addHours( 0), dt1 == dt1.addHours( 0).atUtc() )
    ensureTrue(dt1, "!=", dt1.addHours( 2), dt1 != dt1.addHours( 2).atUtc() )

    // Case 7. Get the duration
    println( dt1.addSeconds(2).durationFrom( dt1 ) )
    println( dt1.addMinutes(2).durationFrom( dt1 ) )
    println( dt1.addHours(2).durationFrom( dt1 ) )
    println( dt1.addDays(2).durationFrom( dt1 ) )
    println( dt1.addMonths(2).periodFrom( dt1 ) )

    //</doc:examples>
    ok()
  }


  def test():Unit = {


    //12 december 2014
    val date3 = LocalDate.of(2014, Month.DECEMBER, 12)
    println("date3: " + date3)

    //22 hour 15 minutes
    val date4 = LocalTime.of(22, 15)
    println("date4: " + date4)

    //parse a string
    val date5 = LocalTime.parse("20:15:30")
    println("date5: " + date5)
  }


  def ensureTrue(date1:DateTime, comp:String, date2:DateTime, result:Boolean) : Unit = {
    println("comparing : " + date1.toString () + " " + comp + " " + date2.toString + " = " + result)
    Ensure.isTrue(result, "Date comparision does not match")
  }
}

File: src/lib/kotlin/slatekit-examples/src/main/kotlin/slatekit/examples/Example_DateTime.kt
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slatekit.examples

//<doc:import_required>
import slatekit.common.DateTime

//</doc:import_required>

//<doc:import_examples>
import slatekit.core.cmds.Cmd
import slatekit.results.Try
import slatekit.results.Success
import slatekit.common.ext.days
import slatekit.common.ext.minutes
import slatekit.common.ext.months
import java.time.ZoneId

//</doc:import_examples>


class Example_DateTime  : Cmd("datetime") {

  override fun executeInternal(args: Array<String>?) : Try<Any>
  {
    //<doc:examples>
    //
    // NOTE:
    // When working with new Java 8 Date/Time, which is a significant
    // improvement over the older mutable Java Date/Time models,
    // there is still some "cognitive overhead" ( IMHO ) in mentally
    // managing the different LocalDateTime, ZonedDateTime and conversion
    // to and from local/zoned functionality
    //
    // DESIGN:
    // This DateTime is a unified DateTime for both LocalDateTime and ZonedDateime
    // that wraps uses a ZonedDateTime internally ( defaulted to local timezone )
    // and is used for representing a DateTime for either Local and/or other Zones.
    // This makes the Java 8 datetime/zones a bit simpler & concise while
    // essentially adding syntactic sugar using Kotlin operators and extension methods
    //
    // IMPORTANT:
    // This does NOT change the functionality of the Java 8 classes at all.
    // It is simply "syntactic sugar" for using the classes.

    // Case 1. Get datetime now, either locally, at utc and other zones
    // These will return a DateTime that wraps a ZonedDateTime.
    println( DateTime.now() )
    println( DateTime.nowUtc() )
    println( DateTime.nowAt("America/New_York"))
    println( DateTime.nowAt("Europe/Athens"))
    println( DateTime.nowAt(ZoneId.of("Europe/Athens")))


    // Case 2: Build datetime explicitly
    println( DateTime.of(2017, 7, 10))
    println( DateTime.of(2017, 7, 10, 11, 30, 0))
    println( DateTime.of(2017, 7, 10, 11, 30, 0, 0, "America/New_York"))
    println( DateTime.of(2017, 7, 10, 11, 30, 0, 0, ZoneId.of("America/New_York")))


    // Case 3. Get datetime fields
    val dt = DateTime.now()
    println( "year  : " + dt.year      )
    println( "month : " + dt.month     )
    println( "day   : " + dt.day       )
    println( "hour  : " + dt.hours     )
    println( "mins  : " + dt.minutes   )
    println( "secs  : " + dt.seconds   )
    println( "nano  : " + dt.nano      )
    println( "zone  : " + dt.zone.id )


    // Case 4: Conversion from now( local ) to utc, specific zone,
    // LocalDate, LocalTime, and LocalDateTime
    val now      = DateTime.now()
    println( now.date()  )
    println( now.time()  )
    println( now.local() )
    println( now.atUtc() )
    println( now.atUtcLocal() )
    println( now.atZone("Europe/Athens") )


    // Case 5: Idiomatic use of Kotlin operators and extension methods
    // This uses the extensions from slatekit.common.ext.IntExtensions
    val later = DateTime.now() + 3.minutes
    println(    DateTime.now() + 3.days            )
    println(    DateTime.now() - 3.minutes         )
    println(    DateTime.now() - 3.months          )
    println(    DateTime.now().secondsTo( later )  )
    println(    DateTime.now().minutesTo( later )  )


    // Case 6. Add time ( just like Java 8 )
    val dt1 = DateTime.now()
    println( dt1.plusYears  (1).toString() )
    println( dt1.plusMonths (1).toString() )
    println( dt1.plusDays   (1).toString() )
    println( dt1.plusHours  (1).toString() )
    println( dt1.plusMinutes(1).toString() )
    println( dt1.plusSeconds(1).toString() )


    // Case 7. Compare
    println( dt1 >   dt1.plusYears  (1) )
    println( dt1 >=  dt1.plusMonths (1) )
    println( dt1 >=  dt1.plusDays   (1) )
    println( dt1 <   dt1.plusHours  (1) )
    println( dt1 <=  dt1.plusMinutes(1) )
    println( dt1 <=  dt1.plusSeconds(1) )


    // Case 8. Get duration (hours,mins,seconds) or period(days,months,years)
    println( dt1.plusSeconds(2).durationFrom( dt1 ) )
    println( dt1.plusMinutes(2).durationFrom( dt1 ) )
    println( dt1.plusHours(2).durationFrom( dt1 ) )
    println( dt1.plusDays(2).periodFrom( dt1 ) )
    println( dt1.plusMonths(2).periodFrom( dt1 ) )
    println( dt1.plusYears(2).periodFrom( dt1 ) )

    //</doc:examples>
    return Success("")
  }
}
